buildmodel/buildassets/buildassets_test.go

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package buildassets
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"testing"
+)
+
+func Test_getVersion(t *testing.T) {
+	tests := []struct {
+		name        string
+		text        string
+		wantVersion string
+	}{
+		{"Ordinary", "go1.17.3", "go1.17.3"},
+		{"TrailingNewline", "go1.17.3\n", "go1.17.3"},
+		{"MultilineFile", "go1.17.3\nMore information", "go1.17.3"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			filePath := path.Join(t.TempDir(), "VERSION")
+			err := ioutil.WriteFile(filePath, []byte(tt.text), os.ModePerm)
```

Comment:
`io/ioutil` is deprecated since 1.16. Better us [os#WriteFile](https://pkg.go.dev/os#WriteFile) unless this has to be compiled with a previous version.

Reply:
Is there a way to trigger diagnostics when using deprecated APIs? Something we can hook into our build to catch it then?

Reply:
staticcheck detects deprecated functions (tracking potentially adding that tool with https://github.com/microsoft/go/issues/78), but these `ioutil` functions aren't actually *marked* as deprecated even though https://go.dev/doc/go1.16 did announce it.

https://go.dev/blog/godoc says:

> Sometimes a struct field, function, type, or even a whole package becomes redundant or unnecessary, but must be kept for compatibility with existing programs. To signal that an identifier should not be used, **add a paragraph to its doc comment that begins with “Deprecated:” followed by some information about the deprecation**.

The comments on these deprecated funcs don't follow that guidance:

```
#> go doc ioutil.WriteFile
package ioutil // import "io/ioutil"

func WriteFile(filename string, data []byte, perm fs.FileMode) error
    WriteFile writes data to a file named by filename. If the file does not
    exist, WriteFile creates it with permissions perm (before umask); otherwise
    WriteFile truncates it before writing, without changing permissions.

    As of Go 1.16, this function simply calls os.WriteFile.
```

Reply:
https://go-review.googlesource.com/c/go/+/284777/ has more detail about why these aren't marked:

> I think it's a little soon to call these deprecated.  In general we use "deprecated" markers to tell people that they need to change their code, but this code continues to work just fine.  Let's let people adjust over a couple of releases and then reconsider.

> I agree with Ian that it is too soon to start flagging these uses to people for conversion. People who are still supporting modules built using Go 1.15 _cannot_ start using these functions, so it would be not just useless advice but bad advice. This does not seem to fit "Deprecated" to me. If we do the new inlining-based go fix then these would be good candidates for that. They still wouldn't be deprecated. To me Deprecated implies an important reason not to use the function. There is no important reason not to use these functions, and there is a good reason to keep using them (Go 1.15 compatibility).

Reply:
Thanks for bringing up the full picture 😄 It is clearly not deprecated, just outdated. I'm not aware of any tool that flags its usage, but I do think it is a good practice to avoid using when possible.